### PR TITLE
[4.0] Dashboards

### PR DESCRIPTION
**Pull request by brianteeman**
Before this PR In the dashboards of joomla 4 we have a situation where sometimes there is a large space between _modules_

This is because the _modules_ generated by the system are in a different div than that of the more traditional modules.

By moving them all into the same div that problem disappears. I can't see any reason for them to be in different divs

### Example Home Before
image

### Example Home After
image

### CODE REVIEW

**Comment by richard67** [comment]
@brianteeman The useless space has gone, but now ordering is different so the update checks are a bit out of focus because you have to scroll down to see them. Beside this I like the new look. Maybe we just can reorder the modules a bit?

**Comment by richard67** [comment]
P.S. And thanks for quickly showing the guys what can be done when just someone does it.

**Comment by brianteeman** [comment]
the ordering is just css - you just have to change the direction

**Comment by richard67** [comment]
@brianteeman "you just have to change the direction" ... "you" means me? Or one? Will you do that? Or shall we test this PR here as is and do it with another future one?

**Comment by brianteeman** [comment]
> do it with another future one?

yes

**Comment by richard67** [comment]
@brianteeman You know who can do that? I am not an (s)css expert.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 2275109a7fca79694f8c12da27a2758a3e03700d<br /><br />Dashboards are not wasting empty space anymore.

Ordering of modules is different now so the update checks are a bit out of focus because you have to scroll down to see them.

But this can be fixed with a future PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on 2275109a7fca79694f8c12da27a2758a3e03700d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by richard67** [comment]
@brianteeman I trust you that ordering can be easily changed.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by brianteeman** [comment]
Removing the rtc and asking for help on the scss (I might be wrong that direction can be changed here)

**Comment by richard67** [comment]
> Removing the rtc and asking for help on the scss (I might be wrong that direction can be changed here)

@C-Lodder Can you advise here?

**Comment by richard67** [comment]
@brianteeman It seems the bot has added back RTC for some reason. not sure if related to my comment - I only commented on HitHub - but there is a coincidence in time.

**Comment by dgrammatiko** [comment]
> the ordering is just css

That would have been the case if Joomla was using the Platform and the CSS Grid Layout instead you're using an extremely opinionated and outdated grid system which doesn't support such things without **editing** the html. All this B/C break just to be in the same position the project was 10 years ago. Sad...

**Comment by brianteeman** [comment]
I am not using anything

If you dont have anything constructive to say ....

**Comment by dgrammatiko** [comment]
I've already said something constructive, use CSS Grid which allows you do what you want. Bootstrap doesn't. Honestly I keep saying this for years...

**Comment by C-Lodder** [comment]
@richard67 Doesn't the ordering derive from the "Ordering" parameter for each module?

**Comment by richard67** [comment]
@alikon It seems that whenever someone comments on this PR, a previously removed RTC label is added back. Have you removed it by changing status in the issue tracker, or just by removing the label on GitHub? (sorry @brianteeman if off topic, let me know when we shall continue to discuss that elsewhere).

**Comment by brianteeman** [comment]
@c-lodder yes it does. I was thinking we can change from column  ordering to row ordering

**Comment by ciar4n** [comment]
The dashboard is using `column-count` so the option does not exist to simply change direction. They will always flow as you would expect paragraphs in a multi-column page.

**Comment by ciar4n** [comment]
CSS grid is not really the solution either as it does not support a true masonry layout. The only CSS solution is probably flex. A bit of code involved but the markup would remain the same.

**Comment by ciar4n** [comment]
Actually, you probably could use CSS grid although it is a bit hacky. Set 2 columns and then use `:nth-child()` to add every second item to the second column.

**Comment by brianteeman** [comment]
flex is exactly what i was thinking of

**Comment by ciar4n** [comment]
The solution with grid and flex would be quite similar (both using `:nth-child()`).

**Comment by ciar4n** [comment]
This should help... https://tobiasahlin.com/blog/masonry-with-css/

**Comment by ciar4n** [comment]
Using CSS grid would be little easier and more versatile. If you wish, this can be merged and I will create a follow-up PR.

**Comment by brianteeman** [comment]
that would be the ideal - thanks

**Comment by richard67** [comment]
@brianteeman Shall I add back RTC?

**Comment by infograf768** [comment]
> Shall I add back RTC?

Imho not until a full solution is proposed.

**Comment by ciar4n** [comment]
> Imho not until a full solution is proposed.

@infograf768 what are we missng?

**Comment by infograf768** [comment]
> @infograf768 what are we missing?

@richard67  explained it already
Updates are now far down and need to scroll.
I will have for example Popular Articles at the same level as Site

Screen Shot 2020-02-04 at 10 03 47

This PR just makes it more difficult to access the important stuff just to take off a bit of free space.

**Comment by ciar4n** [comment]
> Using CSS grid would be little easier and more versatile. If you wish, this can be merged and I will create a follow-up PR.

The solution I offered would move items listed first to the top of the dashboard

**Comment by infograf768** [comment]
> The solution I offered would move items listed first to the top of the dashboard

Great. Then let's close this one and have a single PR doing what it should (and by default) so that we can really test the final result.

**Comment by ciar4n** [comment]
Unfortunately, I'm gonna have to renege on this. I've explored all possible CSS only solution and they all come with limitations that make it unsuitable (eg. require a set height on the container). We had come to the same conclusion with the previous atum template but I had somehow forgotten it was not possible. 

As before, the only possible solutions is to either create 2 positions, one for each column or use javascript.

**Comment by brianteeman** [comment]
> As before, the only possible solutions is to either create 2 positions, one for each column or use javascript.

Or change the ordering in the sql so that a fresh install doesnt have the issue @richard67 raised? will look at that soon

**Comment by brianteeman** [comment]
@ciar4n Thanks for checking etc

**Comment by infograf768** [comment]
OK. In this case and unless someone proposes a useful alternative. I suggest to close this PR.
We can live with a bit of wasted space.

**Comment by infograf768** [comment]
I have tested this item :red_circle: unsuccessfully on 2275109a7fca79694f8c12da27a2758a3e03700d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by Fedik** [comment]
This is kind of [masonry layout](https://css-tricks.com/piecing-together-approaches-for-a-css-masonry-layout/), "row flow" possible only with js

> As before, the only possible solutions is to either create 2 positions, one for each column or use javascript.

@ciar4n  Maybe a little JS does not hurt here :wink:
https://css-tricks.com/piecing-together-approaches-for-a-css-masonry-layout/#article-header-id-5

**Comment by brianteeman** [comment]
@infograf768 please don't be so negative - there are many ways to skin a cat. It is fine to say that you don't agree with the code. It is fine to say you don't see the need. Marking the code as an unsuccessful test on that criteria is not ok. A test should mark if the code does what it says and nothing more. A test is a factual statement and not an opinion. If you don't want to mark something as tested because you don't agree then that is also fine.

**Comment by ciar4n** [comment]
> Or change the ordering in the sql so that a fresh install doesnt have the issue @richard67 raised? will look at that soon

@brianteeman You will have an issue there if a user adds/removes any modules. `column-count` will alter the layout so both columns are equal, potentially moving modules from the top.

@Fedik Yes I think a little JS is perfectly acceptable here. I didn't fully understand the JS involved which is why I didn't attempt it myself.

**Comment by brianteeman** [comment]
@fedik is this something you are able to investigate?

**Comment by Fedik** [comment]
> I didn't fully understand the JS involved which is why I didn't attempt it myself.

Need to calculate `grid-row-end` for every cell, somehow
Here more detailed explanation https://medium.com/@andybarefoot/a-masonry-style-layout-using-css-grid-8c663d355ebb

> is this something you are able to investigate?

I can try to check when will get some time

**Comment by brianteeman** [comment]
@Fedik Thanks - in the meantime I will read that blog post you linked to

**Comment by brianteeman** [comment]
Is this acceptable?

https://codepen.io/brianteeman/pen/OJVJLLG

There is no change to the markup. Just css and js

**Comment by infograf768** [comment]
Looks good (I guess the patch would just change the uneeded space). Would we still be able to modify the order of modules?

**Comment by brianteeman** [comment]
yes because this does not touch the markup

**Comment by brianteeman** [comment]
And of course with this PR in general you are able to change the order of ALL the modules on the home dashboard in one group whereas before this PR you could only change them within two groups

**Comment by Fedik** [comment]
Ignore Hound, the JavaScript need to be optimized a bit, 
I try look in the evening

**Comment by brianteeman** [comment]
ok thanks

**Comment by dgrammatiko** [comment]
@Fedik please use the native resizeObserver (if supported) or fallback to resize (either way you need to debounce a bit)

**Comment by Fedik** [comment]
hm, I try one more time :smile:

**Comment by infograf768** [comment]
Hound should be impeached...

File: build/media_source/com_cpanel/js/admin-cpanel-default.es6.js [hunk 1]
```diff
@@ -73,4 +73,47 @@
...
+      const gridStyle = window.getComputedStyle(MasonryLayout.$gridBox);
+      const gridAutoRows = parseInt(gridStyle.getPropertyValue('grid-auto-rows'), 10) || 0;
+      const gridRowGap = parseInt(gridStyle.getPropertyValue('grid-row-gap'), 10) || 10;
+
+      $gridCells.forEach(($cell) => {
+        MasonryLayout.resizeGridItem($cell, gridAutoRows, gridRowGap);
+      });
+    },
+
+    initialise () {
```
[the remark this answers is not in the document]
Comment (Fedik) [reply build/media_source/com_cpanel/js/admin-cpanel-default.es6.js]: this annoying, first he want space, now he do not want space

[the remark this answers is not in the document]
Comment (richard67) [reply build/media_source/com_cpanel/js/admin-cpanel-default.es6.js]: Maybe the Hound is a Poodle now? ;-)

[the remark this answers is not in the document]
Comment (richard67) [reply build/media_source/com_cpanel/js/admin-cpanel-default.es6.js]: ah i know. it wants space before `{`, but it doesn't want space before `()`.

**Comment by wilsonge** [comment]
> Hound should be impeached...

Unfortunately hound has a majority in the senior chamber of release leads ;)

**Comment by richard67** [comment]
@wilsonge After the Hound was happy, AppVeyor got mad. Chocolatey faild to install php or something like that. Seems not to be related to this PR. Maybe the Hound has eaten all the Chocolate(y)?

**Comment by brianteeman** [comment]
appveyor has that problem a lot!!

**Comment by richard67** [comment]
If those test would work reliably we could say we don't give RTC when tests bad .. but like it is with that lotto game we can't do that, it would block so many PRs. Sometimes it helps to retrigger the CI tests by pulling in the base branch, but if nothing happens on the base branch that doesn't help. Only way then is to commit a dummy change. It's not really fun to work like that :-(

**Comment by wilsonge** [comment]
PHP Is timing out installing :( We can probably bump that timeout up

**Comment by richard67** [comment]
I've just tested but think it could be optimized. CSS or JS experts, whatever needed here, please check and advise.

When having my browser window in a size similar as shown in all the screnshot above, I see now following:

test-pr27777_1

I.e. buttons are too big, all is in 1 column, so I have to scroll down to see the update checks.

When I enlarge browser width to covering almost the full screen (which is 1920px wide!!!) the buttons look ok, i.e. buttons are like before the PR, and I can see all update checks with 1 look:

test-pr27777_2

So it seems that below a certain window width it switches to 1 column, but this change happens too early, i.e. when browser window is still big enough to show it nice in 2 columns. It should change on smaller width than like it is now.

Beside this, maybe it would be better to exchange position of "Site" and "Update Checks" (ordering of the modules) so the update checks always come first and so are on top when 1 column only?

**Comment by richard67** [comment]
So PR works, but it could be optimized a bit as described in my previous comment.

**Comment by brianteeman** [comment]
I agree

**Comment by Fedik** [comment]
> So it seems that below a certain window width it switches to 1 column

this with it:
`grid-template-columns: repeat(auto-fill, minmax(700px,1fr));`
if the cell cannot fit 700px it goes 100% (1fr)

I found 700px work good for my screen and mobile, but I totally forgot about "middle sizes" 

Need try to play with it a bit, maybe something down to 560

**Comment by brianteeman** [comment]
Yeah. That's the line to play with. And don't forget it's different on the system and help pages

**Comment by richard67** [comment]
@Fedik It seems your mobile phone display is bigger than the parking space I have for my car 😄

**Comment by Fedik** [comment]
or just change it to more simle:
for large:
`grid-template-columns: 1fr 1fr;`
for small:
`grid-template-columns: 1fr;`

**Comment by richard67** [comment]
Hey, seems to be cool now, I like the result:
test_pr-27777_1

test_pr-27777_2

test_pr-27777_3

test_pr-27777_4

Good team work!

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 5ff0fe7a2c91299319986a8d8ef92113c88a4e5f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by richard67** [comment]
Really nice. We can make future PR to move upate checks to the top by module order. But so or so I love how the modules and the buttons mode and scale now when reducing display width.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 5ff0fe7a2c91299319986a8d8ef92113c88a4e5f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27777">issues.joomla.org/tracker/joomla-cms/27777</a>.</sub>

**Comment by richard67** [comment]
@brianteeman Shall we wait a bit for the requested reviews by @rdeutz and @wilsonge before we merge?

Update: I have no problem to merge it, it is awesome, but you had requested those reviews so I thought I ask if you are still waiting for them.

**Comment by richard67** [comment]
Of course if @wilsonge merges this it will be as good as review. ;-)

**Review by wilsonge — APPROVED** [review]

**Comment by brianteeman** [comment]
Thanks. Making the impossible possible is nice

**Pull request merged**